// @signalapp/libsignal-client/dist/net/KeyTransparency.js
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import * as Native from '../Native.js';
/**
 *  Mode of the monitor operation.
 *
 *  If the newer version of account data is found in the key transparency
 *  log, self-monitor will terminate with an error, but monitor for other
 *  account will fall back to a full search and update the locally stored
 *  data.
 */
export var MonitorMode;
(function (MonitorMode) {
    MonitorMode[MonitorMode["Self"] = 0] = "Self";
    MonitorMode[MonitorMode["Other"] = 1] = "Other";
})(MonitorMode || (MonitorMode = {}));
export class ClientImpl {
    constructor(asyncContext, chatService, env) {
        this.asyncContext = asyncContext;
        this.chatService = chatService;
        this.env = env;
    }
    async search(request, store, options) {
        const distinguished = await this._getLatestDistinguished(store, options ?? {});
        const { abortSignal } = options ?? {};
        const { aciInfo: { aci, identityKey: aciIdentityKey }, e164Info, usernameHash, } = request;
        const { e164, unidentifiedAccessKey } = e164Info ?? {
            e164: null,
            unidentifiedAccessKey: null,
        };
        const accountData = await this.asyncContext.makeCancellable(abortSignal, Native.KeyTransparency_Search(this.asyncContext, this.env, this.chatService, aci.getServiceIdFixedWidthBinary(), aciIdentityKey, e164, unidentifiedAccessKey, usernameHash ?? null, await store.getAccountData(aci), distinguished));
        await store.setAccountData(aci, accountData);
    }
    async monitor(request, store, options) {
        const distinguished = await this._getLatestDistinguished(store, options ?? {});
        const { abortSignal } = options ?? {};
        const { aciInfo: { aci, identityKey: aciIdentityKey }, e164Info, usernameHash, mode, } = request;
        const { e164, unidentifiedAccessKey } = e164Info ?? {
            e164: null,
            unidentifiedAccessKey: null,
        };
        const accountData = await this.asyncContext.makeCancellable(abortSignal, Native.KeyTransparency_Monitor(this.asyncContext, this.env, this.chatService, aci.getServiceIdFixedWidthBinary(), aciIdentityKey, e164, unidentifiedAccessKey, usernameHash ?? null, await store.getAccountData(aci), distinguished, mode === MonitorMode.Self));
        await store.setAccountData(aci, accountData);
    }
    async updateDistinguished(store, { abortSignal }) {
        const bytes = await this.asyncContext.makeCancellable(abortSignal, Native.KeyTransparency_Distinguished(this.asyncContext, this.env, this.chatService, await store.getLastDistinguishedTreeHead()));
        await store.setLastDistinguishedTreeHead(bytes);
        return bytes;
    }
    async _getLatestDistinguished(store, options) {
        return ((await store.getLastDistinguishedTreeHead()) ??
            (await this.updateDistinguished(store, options)));
    }
}
