// @signalapp/libsignal-client/dist/index.js
//
// Copyright 2020-2022 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//
import { Buffer } from 'node:buffer';
import * as uuid from 'uuid';
import * as Errors from './Errors.js';
export * from './Errors.js';
import { Aci, ProtocolAddress, ServiceId } from './Address.js';
export * from './Address.js';
import { PrivateKey, PublicKey } from './EcKeys.js';
export * from './EcKeys.js';
export * from './uuid.js';
export * as usernames from './usernames.js';
export * as io from './io.js';
export * as Net from './net.js';
export * as Mp4Sanitizer from './Mp4Sanitizer.js';
export * as WebpSanitizer from './WebpSanitizer.js';
import * as Native from './Native.js';
Native.registerErrors(Errors);
// These enums must be kept in sync with their Rust counterparts.
export var CiphertextMessageType;
(function (CiphertextMessageType) {
    CiphertextMessageType[CiphertextMessageType["Whisper"] = 2] = "Whisper";
    CiphertextMessageType[CiphertextMessageType["PreKey"] = 3] = "PreKey";
    CiphertextMessageType[CiphertextMessageType["SenderKey"] = 7] = "SenderKey";
    CiphertextMessageType[CiphertextMessageType["Plaintext"] = 8] = "Plaintext";
})(CiphertextMessageType || (CiphertextMessageType = {}));
export var Direction;
(function (Direction) {
    Direction[Direction["Sending"] = 0] = "Sending";
    Direction[Direction["Receiving"] = 1] = "Receiving";
})(Direction || (Direction = {}));
// This enum must be kept in sync with sealed_sender.proto.
export var ContentHint;
(function (ContentHint) {
    ContentHint[ContentHint["Default"] = 0] = "Default";
    ContentHint[ContentHint["Resendable"] = 1] = "Resendable";
    ContentHint[ContentHint["Implicit"] = 2] = "Implicit";
})(ContentHint || (ContentHint = {}));
export function hkdf(outputLength, keyMaterial, label, salt) {
    return Native.HKDF_DeriveSecrets(outputLength, keyMaterial, label, salt);
}
export class ScannableFingerprint {
    constructor(scannable) {
        this.scannable = scannable;
    }
    static _fromBuffer(scannable) {
        return new ScannableFingerprint(scannable);
    }
    compare(other) {
        return Native.ScannableFingerprint_Compare(this.scannable, other.scannable);
    }
    toBuffer() {
        return this.scannable;
    }
}
export class DisplayableFingerprint {
    constructor(display) {
        this.display = display;
    }
    static _fromString(display) {
        return new DisplayableFingerprint(display);
    }
    toString() {
        return this.display;
    }
}
export class Fingerprint {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static new(iterations, version, localIdentifier, localKey, remoteIdentifier, remoteKey) {
        return new Fingerprint(Native.Fingerprint_New(iterations, version, localIdentifier, localKey, remoteIdentifier, remoteKey));
    }
    displayableFingerprint() {
        return DisplayableFingerprint._fromString(Native.Fingerprint_DisplayString(this));
    }
    scannableFingerprint() {
        return ScannableFingerprint._fromBuffer(Native.Fingerprint_ScannableEncoding(this));
    }
}
/**
 * Implements the <a href="https://en.wikipedia.org/wiki/AES-GCM-SIV">AES-256-GCM-SIV</a>
 * authenticated stream cipher with a 12-byte nonce.
 *
 * AES-GCM-SIV is a multi-pass algorithm (to generate the "synthetic initialization vector"), so
 * this API does not expose a streaming form.
 */
export class Aes256GcmSiv {
    constructor(key) {
        this._nativeHandle = Native.Aes256GcmSiv_New(key);
    }
    static new(key) {
        return new Aes256GcmSiv(key);
    }
    /**
     * Encrypts the given plaintext using the given nonce, and authenticating the ciphertext and given
     * associated data.
     *
     * The associated data is not included in the ciphertext; instead, it's expected to match between
     * the encrypter and decrypter. If you don't need any extra data, pass an empty array.
     *
     * @returns The encrypted data, including an appended 16-byte authentication tag.
     */
    encrypt(message, nonce, associatedData) {
        return Native.Aes256GcmSiv_Encrypt(this, message, nonce, associatedData);
    }
    /**
     * Decrypts the given ciphertext using the given nonce, and authenticating the ciphertext and given
     * associated data.
     *
     * The associated data is not included in the ciphertext; instead, it's expected to match between
     * the encrypter and decrypter.
     *
     * @returns The decrypted data
     */
    decrypt(message, nonce, associatedData) {
        return Native.Aes256GcmSiv_Decrypt(this, message, nonce, associatedData);
    }
}
export class KEMPublicKey {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _fromNativeHandle(handle) {
        return new KEMPublicKey(handle);
    }
    static deserialize(buf) {
        return new KEMPublicKey(Native.KyberPublicKey_Deserialize(buf));
    }
    serialize() {
        return Native.KyberPublicKey_Serialize(this);
    }
}
export class KEMSecretKey {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _fromNativeHandle(handle) {
        return new KEMSecretKey(handle);
    }
    static deserialize(buf) {
        return new KEMSecretKey(Native.KyberSecretKey_Deserialize(buf));
    }
    serialize() {
        return Native.KyberSecretKey_Serialize(this);
    }
}
export class KEMKeyPair {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _fromNativeHandle(handle) {
        return new KEMKeyPair(handle);
    }
    static generate() {
        return new KEMKeyPair(Native.KyberKeyPair_Generate());
    }
    getPublicKey() {
        return KEMPublicKey._fromNativeHandle(Native.KyberKeyPair_GetPublicKey(this));
    }
    getSecretKey() {
        return KEMSecretKey._fromNativeHandle(Native.KyberKeyPair_GetSecretKey(this));
    }
}
export class PreKeyBundle {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static new(registration_id, device_id, prekey_id, prekey, signed_prekey_id, signed_prekey, signed_prekey_signature, identity_key, kyber_prekey_id, kyber_prekey, kyber_prekey_signature) {
        return new PreKeyBundle(Native.PreKeyBundle_New(registration_id, device_id, prekey_id, prekey, signed_prekey_id, signed_prekey, signed_prekey_signature, identity_key, kyber_prekey_id, kyber_prekey, kyber_prekey_signature));
    }
    deviceId() {
        return Native.PreKeyBundle_GetDeviceId(this);
    }
    identityKey() {
        return PublicKey._fromNativeHandle(Native.PreKeyBundle_GetIdentityKey(this));
    }
    preKeyId() {
        return Native.PreKeyBundle_GetPreKeyId(this);
    }
    preKeyPublic() {
        const handle = Native.PreKeyBundle_GetPreKeyPublic(this);
        if (handle == null) {
            return null;
        }
        else {
            return PublicKey._fromNativeHandle(handle);
        }
    }
    registrationId() {
        return Native.PreKeyBundle_GetRegistrationId(this);
    }
    signedPreKeyId() {
        return Native.PreKeyBundle_GetSignedPreKeyId(this);
    }
    signedPreKeyPublic() {
        return PublicKey._fromNativeHandle(Native.PreKeyBundle_GetSignedPreKeyPublic(this));
    }
    signedPreKeySignature() {
        return Native.PreKeyBundle_GetSignedPreKeySignature(this);
    }
    kyberPreKeyId() {
        return Native.PreKeyBundle_GetKyberPreKeyId(this);
    }
    kyberPreKeyPublic() {
        return KEMPublicKey._fromNativeHandle(Native.PreKeyBundle_GetKyberPreKeyPublic(this));
    }
    kyberPreKeySignature() {
        return Native.PreKeyBundle_GetKyberPreKeySignature(this);
    }
}
export class PreKeyRecord {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new PreKeyRecord(nativeHandle);
    }
    static new(id, pubKey, privKey) {
        return new PreKeyRecord(Native.PreKeyRecord_New(id, pubKey, privKey));
    }
    static deserialize(buffer) {
        return new PreKeyRecord(Native.PreKeyRecord_Deserialize(buffer));
    }
    id() {
        return Native.PreKeyRecord_GetId(this);
    }
    privateKey() {
        return PrivateKey._fromNativeHandle(Native.PreKeyRecord_GetPrivateKey(this));
    }
    publicKey() {
        return PublicKey._fromNativeHandle(Native.PreKeyRecord_GetPublicKey(this));
    }
    serialize() {
        return Native.PreKeyRecord_Serialize(this);
    }
}
export class SignedPreKeyRecord {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new SignedPreKeyRecord(nativeHandle);
    }
    static new(id, timestamp, pubKey, privKey, signature) {
        return new SignedPreKeyRecord(Native.SignedPreKeyRecord_New(id, timestamp, pubKey, privKey, signature));
    }
    static deserialize(buffer) {
        return new SignedPreKeyRecord(Native.SignedPreKeyRecord_Deserialize(buffer));
    }
    id() {
        return Native.SignedPreKeyRecord_GetId(this);
    }
    privateKey() {
        return PrivateKey._fromNativeHandle(Native.SignedPreKeyRecord_GetPrivateKey(this));
    }
    publicKey() {
        return PublicKey._fromNativeHandle(Native.SignedPreKeyRecord_GetPublicKey(this));
    }
    serialize() {
        return Native.SignedPreKeyRecord_Serialize(this);
    }
    signature() {
        return Native.SignedPreKeyRecord_GetSignature(this);
    }
    timestamp() {
        return Native.SignedPreKeyRecord_GetTimestamp(this);
    }
}
export class KyberPreKeyRecord {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new KyberPreKeyRecord(nativeHandle);
    }
    static new(id, timestamp, keyPair, signature) {
        return new KyberPreKeyRecord(Native.KyberPreKeyRecord_New(id, timestamp, keyPair, signature));
    }
    serialize() {
        return Native.KyberPreKeyRecord_Serialize(this);
    }
    static deserialize(buffer) {
        return new KyberPreKeyRecord(Native.KyberPreKeyRecord_Deserialize(buffer));
    }
    id() {
        return Native.KyberPreKeyRecord_GetId(this);
    }
    keyPair() {
        return KEMKeyPair._fromNativeHandle(Native.KyberPreKeyRecord_GetKeyPair(this));
    }
    publicKey() {
        return KEMPublicKey._fromNativeHandle(Native.KyberPreKeyRecord_GetPublicKey(this));
    }
    secretKey() {
        return KEMSecretKey._fromNativeHandle(Native.KyberPreKeyRecord_GetSecretKey(this));
    }
    signature() {
        return Native.KyberPreKeyRecord_GetSignature(this);
    }
    timestamp() {
        return Native.KyberPreKeyRecord_GetTimestamp(this);
    }
}
export class SignalMessage {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _new(messageVersion, macKey, senderRatchetKey, counter, previousCounter, ciphertext, senderIdentityKey, receiverIdentityKey, pqRatchet) {
        return new SignalMessage(Native.SignalMessage_New(messageVersion, macKey, senderRatchetKey, counter, previousCounter, ciphertext, senderIdentityKey, receiverIdentityKey, pqRatchet));
    }
    static deserialize(buffer) {
        return new SignalMessage(Native.SignalMessage_Deserialize(buffer));
    }
    body() {
        return Native.SignalMessage_GetBody(this);
    }
    pqRatchet() {
        return Native.SignalMessage_GetPqRatchet(this);
    }
    counter() {
        return Native.SignalMessage_GetCounter(this);
    }
    messageVersion() {
        return Native.SignalMessage_GetMessageVersion(this);
    }
    serialize() {
        return Native.SignalMessage_GetSerialized(this);
    }
    verifyMac(senderIdentityKey, recevierIdentityKey, macKey) {
        return Native.SignalMessage_VerifyMac(this, senderIdentityKey, recevierIdentityKey, macKey);
    }
}
export class PreKeySignalMessage {
    constructor(handle) {
        this._nativeHandle = handle;
    }
    static _new(messageVersion, registrationId, preKeyId, signedPreKeyId, baseKey, identityKey, signalMessage) {
        return new PreKeySignalMessage(Native.PreKeySignalMessage_New(messageVersion, registrationId, preKeyId, signedPreKeyId, baseKey, identityKey, signalMessage));
    }
    static deserialize(buffer) {
        return new PreKeySignalMessage(Native.PreKeySignalMessage_Deserialize(buffer));
    }
    preKeyId() {
        return Native.PreKeySignalMessage_GetPreKeyId(this);
    }
    registrationId() {
        return Native.PreKeySignalMessage_GetRegistrationId(this);
    }
    signedPreKeyId() {
        return Native.PreKeySignalMessage_GetSignedPreKeyId(this);
    }
    version() {
        return Native.PreKeySignalMessage_GetVersion(this);
    }
    serialize() {
        return Native.PreKeySignalMessage_Serialize(this);
    }
}
export class SessionRecord {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new SessionRecord(nativeHandle);
    }
    static deserialize(buffer) {
        return new SessionRecord(Native.SessionRecord_Deserialize(buffer));
    }
    serialize() {
        return Native.SessionRecord_Serialize(this);
    }
    archiveCurrentState() {
        Native.SessionRecord_ArchiveCurrentState(this);
    }
    localRegistrationId() {
        return Native.SessionRecord_GetLocalRegistrationId(this);
    }
    remoteRegistrationId() {
        return Native.SessionRecord_GetRemoteRegistrationId(this);
    }
    /**
     * Returns whether the current session can be used to send messages.
     *
     * If there is no current session, returns false.
     */
    hasCurrentState(now = new Date()) {
        return Native.SessionRecord_HasUsableSenderChain(this, now.getTime());
    }
    currentRatchetKeyMatches(key) {
        return Native.SessionRecord_CurrentRatchetKeyMatches(this, key);
    }
}
export class ServerCertificate {
    static _fromNativeHandle(nativeHandle) {
        return new ServerCertificate(nativeHandle);
    }
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static new(keyId, serverKey, trustRoot) {
        return new ServerCertificate(Native.ServerCertificate_New(keyId, serverKey, trustRoot));
    }
    static deserialize(buffer) {
        return new ServerCertificate(Native.ServerCertificate_Deserialize(buffer));
    }
    certificateData() {
        return Native.ServerCertificate_GetCertificate(this);
    }
    key() {
        return PublicKey._fromNativeHandle(Native.ServerCertificate_GetKey(this));
    }
    keyId() {
        return Native.ServerCertificate_GetKeyId(this);
    }
    serialize() {
        return Native.ServerCertificate_GetSerialized(this);
    }
    signature() {
        return Native.ServerCertificate_GetSignature(this);
    }
}
export class SenderKeyRecord {
    static _fromNativeHandle(nativeHandle) {
        return new SenderKeyRecord(nativeHandle);
    }
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static deserialize(buffer) {
        return new SenderKeyRecord(Native.SenderKeyRecord_Deserialize(buffer));
    }
    serialize() {
        return Native.SenderKeyRecord_Serialize(this);
    }
}
export class SenderCertificate {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new SenderCertificate(nativeHandle);
    }
    static new(senderUuid, senderE164, senderDeviceId, senderKey, expiration, signerCert, signerKey) {
        if (typeof senderUuid !== 'string') {
            senderUuid = senderUuid.getServiceIdString();
        }
        return new SenderCertificate(Native.SenderCertificate_New(senderUuid, senderE164, senderDeviceId, senderKey, expiration, signerCert, signerKey));
    }
    static deserialize(buffer) {
        return new SenderCertificate(Native.SenderCertificate_Deserialize(buffer));
    }
    serialize() {
        return Native.SenderCertificate_GetSerialized(this);
    }
    certificate() {
        return Native.SenderCertificate_GetCertificate(this);
    }
    expiration() {
        return Native.SenderCertificate_GetExpiration(this);
    }
    key() {
        return PublicKey._fromNativeHandle(Native.SenderCertificate_GetKey(this));
    }
    senderE164() {
        return Native.SenderCertificate_GetSenderE164(this);
    }
    senderUuid() {
        return Native.SenderCertificate_GetSenderUuid(this);
    }
    /**
     * Returns an ACI if the sender is a valid UUID, `null` otherwise.
     *
     * In a future release SenderCertificate will *only* support ACIs.
     */
    senderAci() {
        try {
            return Aci.parseFromServiceIdString(this.senderUuid());
        }
        catch {
            return null;
        }
    }
    senderDeviceId() {
        return Native.SenderCertificate_GetDeviceId(this);
    }
    serverCertificate() {
        return ServerCertificate._fromNativeHandle(Native.SenderCertificate_GetServerCertificate(this));
    }
    signature() {
        return Native.SenderCertificate_GetSignature(this);
    }
    /**
     * Validates `this` against the given trust root at the given current time.
     *
     * @see validateWithTrustRoots
     */
    validate(trustRoot, time) {
        return Native.SenderCertificate_Validate(this, [trustRoot], time);
    }
    /**
     * Validates `this` against the given trust roots at the given current time.
     *
     * Checks the certificate against each key in `trustRoots` in constant time (that is, no result
     * is produced until every key is checked), making sure **one** of them has signed its embedded
     * server certificate. The `time` parameter is compared numerically against ``expiration``, and
     * is not required to use any specific units, but Signal uses milliseconds since 1970.
     */
    validateWithTrustRoots(trustRoots, time) {
        return Native.SenderCertificate_Validate(this, trustRoots, time);
    }
}
function bridgeSenderKeyStore(store) {
    return {
        async storeSenderKey(sender, distributionId, record) {
            return store.saveSenderKey(ProtocolAddress._fromNativeHandle(sender), uuid.stringify(distributionId), SenderKeyRecord._fromNativeHandle(record));
        },
        async loadSenderKey(sender, distributionId) {
            const sk = await store.getSenderKey(ProtocolAddress._fromNativeHandle(sender), uuid.stringify(distributionId));
            return sk ? sk._nativeHandle : null;
        },
    };
}
export class SenderKeyDistributionMessage {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static async create(sender, distributionId, store) {
        const handle = await Native.SenderKeyDistributionMessage_Create(sender, uuid.parse(distributionId), bridgeSenderKeyStore(store));
        return new SenderKeyDistributionMessage(handle);
    }
    static _new(messageVersion, distributionId, chainId, iteration, chainKey, pk) {
        return new SenderKeyDistributionMessage(Native.SenderKeyDistributionMessage_New(messageVersion, uuid.parse(distributionId), chainId, iteration, chainKey, pk));
    }
    static deserialize(buffer) {
        return new SenderKeyDistributionMessage(Native.SenderKeyDistributionMessage_Deserialize(buffer));
    }
    serialize() {
        return Native.SenderKeyDistributionMessage_Serialize(this);
    }
    chainKey() {
        return Native.SenderKeyDistributionMessage_GetChainKey(this);
    }
    iteration() {
        return Native.SenderKeyDistributionMessage_GetIteration(this);
    }
    chainId() {
        return Native.SenderKeyDistributionMessage_GetChainId(this);
    }
    distributionId() {
        return uuid.stringify(Native.SenderKeyDistributionMessage_GetDistributionId(this));
    }
}
export async function processSenderKeyDistributionMessage(sender, message, store) {
    await Native.SenderKeyDistributionMessage_Process(sender, message, bridgeSenderKeyStore(store));
}
export class SenderKeyMessage {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _new(messageVersion, distributionId, chainId, iteration, ciphertext, pk) {
        return new SenderKeyMessage(Native.SenderKeyMessage_New(messageVersion, uuid.parse(distributionId), chainId, iteration, ciphertext, pk));
    }
    static deserialize(buffer) {
        return new SenderKeyMessage(Native.SenderKeyMessage_Deserialize(buffer));
    }
    serialize() {
        return Native.SenderKeyMessage_Serialize(this);
    }
    ciphertext() {
        return Native.SenderKeyMessage_GetCipherText(this);
    }
    iteration() {
        return Native.SenderKeyMessage_GetIteration(this);
    }
    chainId() {
        return Native.SenderKeyMessage_GetChainId(this);
    }
    distributionId() {
        return uuid.stringify(Native.SenderKeyMessage_GetDistributionId(this));
    }
    verifySignature(key) {
        return Native.SenderKeyMessage_VerifySignature(this, key);
    }
}
export class UnidentifiedSenderMessageContent {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new UnidentifiedSenderMessageContent(nativeHandle);
    }
    static new(message, sender, contentHint, groupId) {
        return new UnidentifiedSenderMessageContent(Native.UnidentifiedSenderMessageContent_New(message, sender, contentHint, groupId));
    }
    static deserialize(buffer) {
        return new UnidentifiedSenderMessageContent(Native.UnidentifiedSenderMessageContent_Deserialize(buffer));
    }
    serialize() {
        return Native.UnidentifiedSenderMessageContent_Serialize(this);
    }
    contents() {
        return Native.UnidentifiedSenderMessageContent_GetContents(this);
    }
    msgType() {
        return Native.UnidentifiedSenderMessageContent_GetMsgType(this);
    }
    senderCertificate() {
        return SenderCertificate._fromNativeHandle(Native.UnidentifiedSenderMessageContent_GetSenderCert(this));
    }
    contentHint() {
        return Native.UnidentifiedSenderMessageContent_GetContentHint(this);
    }
    groupId() {
        return Native.UnidentifiedSenderMessageContent_GetGroupId(this);
    }
}
export class SessionStore {
}
export var IdentityChange;
(function (IdentityChange) {
    // This must be kept in sync with the Rust enum of the same name.
    IdentityChange[IdentityChange["NewOrUnchanged"] = 0] = "NewOrUnchanged";
    IdentityChange[IdentityChange["ReplacedExisting"] = 1] = "ReplacedExisting";
})(IdentityChange || (IdentityChange = {}));
export class IdentityKeyStore {
    async _getIdentityKey() {
        const key = await this.getIdentityKey();
        return key._nativeHandle;
    }
    async _getLocalRegistrationId() {
        return this.getLocalRegistrationId();
    }
    async _saveIdentity(name, key) {
        return this.saveIdentity(ProtocolAddress._fromNativeHandle(name), PublicKey._fromNativeHandle(key));
    }
    async _isTrustedIdentity(name, key, sending) {
        const direction = sending ? Direction.Sending : Direction.Receiving;
        return this.isTrustedIdentity(ProtocolAddress._fromNativeHandle(name), PublicKey._fromNativeHandle(key), direction);
    }
    async _getIdentity(name) {
        const key = await this.getIdentity(ProtocolAddress._fromNativeHandle(name));
        if (key == null) {
            return Promise.resolve(null);
        }
        else {
            return key._nativeHandle;
        }
    }
}
export class PreKeyStore {
}
export class SignedPreKeyStore {
}
export class KyberPreKeyStore {
}
export class SenderKeyStore {
}
export async function groupEncrypt(sender, distributionId, store, message) {
    return CiphertextMessage._fromNativeHandle(await Native.GroupCipher_EncryptMessage(sender, uuid.parse(distributionId), message, bridgeSenderKeyStore(store)));
}
export async function groupDecrypt(sender, store, message) {
    return Native.GroupCipher_DecryptMessage(sender, message, bridgeSenderKeyStore(store));
}
export class SealedSenderDecryptionResult {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new SealedSenderDecryptionResult(nativeHandle);
    }
    message() {
        return Native.SealedSenderDecryptionResult_Message(this);
    }
    senderE164() {
        return Native.SealedSenderDecryptionResult_GetSenderE164(this);
    }
    senderUuid() {
        return Native.SealedSenderDecryptionResult_GetSenderUuid(this);
    }
    /**
     * Returns an ACI if the sender is a valid UUID, `null` otherwise.
     *
     * In a future release SenderCertificate will *only* support ACIs.
     */
    senderAci() {
        try {
            return Aci.parseFromServiceIdString(this.senderUuid());
        }
        catch {
            return null;
        }
    }
    deviceId() {
        return Native.SealedSenderDecryptionResult_GetDeviceId(this);
    }
}
export class CiphertextMessage {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new CiphertextMessage(nativeHandle);
    }
    static from(message) {
        return message.asCiphertextMessage();
    }
    serialize() {
        return Native.CiphertextMessage_Serialize(this);
    }
    type() {
        return Native.CiphertextMessage_Type(this);
    }
}
export class PlaintextContent {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static deserialize(buffer) {
        return new PlaintextContent(Native.PlaintextContent_Deserialize(buffer));
    }
    static from(message) {
        return new PlaintextContent(Native.PlaintextContent_FromDecryptionErrorMessage(message));
    }
    serialize() {
        return Native.PlaintextContent_Serialize(this);
    }
    body() {
        return Native.PlaintextContent_GetBody(this);
    }
    asCiphertextMessage() {
        return CiphertextMessage._fromNativeHandle(Native.CiphertextMessage_FromPlaintextContent(this));
    }
}
export class DecryptionErrorMessage {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static _fromNativeHandle(nativeHandle) {
        return new DecryptionErrorMessage(nativeHandle);
    }
    static forOriginal(bytes, type, timestamp, originalSenderDeviceId) {
        return new DecryptionErrorMessage(Native.DecryptionErrorMessage_ForOriginalMessage(bytes, type, timestamp, originalSenderDeviceId));
    }
    static deserialize(buffer) {
        return new DecryptionErrorMessage(Native.DecryptionErrorMessage_Deserialize(buffer));
    }
    static extractFromSerializedBody(buffer) {
        return new DecryptionErrorMessage(Native.DecryptionErrorMessage_ExtractFromSerializedContent(buffer));
    }
    serialize() {
        return Native.DecryptionErrorMessage_Serialize(this);
    }
    timestamp() {
        return Native.DecryptionErrorMessage_GetTimestamp(this);
    }
    deviceId() {
        return Native.DecryptionErrorMessage_GetDeviceId(this);
    }
    ratchetKey() {
        const keyHandle = Native.DecryptionErrorMessage_GetRatchetKey(this);
        if (keyHandle) {
            return PublicKey._fromNativeHandle(keyHandle);
        }
        else {
            return undefined;
        }
    }
}
function bridgeSessionStore(store) {
    return {
        async storeSession(rawAddress, record) {
            return store.saveSession(ProtocolAddress._fromNativeHandle(rawAddress), SessionRecord._fromNativeHandle(record));
        },
        async loadSession(rawAddress) {
            const pk = await store.getSession(ProtocolAddress._fromNativeHandle(rawAddress));
            return pk ? pk._nativeHandle : null;
        },
    };
}
export function processPreKeyBundle(bundle, address, sessionStore, identityStore, now = new Date()) {
    return Native.SessionBuilder_ProcessPreKeyBundle(bundle, address, bridgeSessionStore(sessionStore), identityStore, now.getTime());
}
export async function signalEncrypt(message, address, sessionStore, identityStore, now = new Date()) {
    return CiphertextMessage._fromNativeHandle(await Native.SessionCipher_EncryptMessage(message, address, bridgeSessionStore(sessionStore), identityStore, now.getTime()));
}
export function signalDecrypt(message, address, sessionStore, identityStore) {
    return Native.SessionCipher_DecryptSignalMessage(message, address, bridgeSessionStore(sessionStore), identityStore);
}
function bridgePreKeyStore(store) {
    return {
        async storePreKey(id, record) {
            return store.savePreKey(id, PreKeyRecord._fromNativeHandle(record));
        },
        async loadPreKey(id) {
            const pk = await store.getPreKey(id);
            return pk._nativeHandle;
        },
        async removePreKey(id) {
            return store.removePreKey(id);
        },
    };
}
function bridgeSignedPreKeyStore(store) {
    return {
        async storeSignedPreKey(id, record) {
            return store.saveSignedPreKey(id, SignedPreKeyRecord._fromNativeHandle(record));
        },
        async loadSignedPreKey(id) {
            const pk = await store.getSignedPreKey(id);
            return pk._nativeHandle;
        },
    };
}
function bridgeKyberPreKeyStore(store) {
    return {
        async storeKyberPreKey(id, record) {
            return store.saveKyberPreKey(id, KyberPreKeyRecord._fromNativeHandle(record));
        },
        async loadKyberPreKey(id) {
            const pk = await store.getKyberPreKey(id);
            return pk._nativeHandle;
        },
        async markKyberPreKeyUsed(id, ecPrekeyId, baseKey) {
            return store.markKyberPreKeyUsed(id, ecPrekeyId, PublicKey._fromNativeHandle(baseKey));
        },
    };
}
export function signalDecryptPreKey(message, address, sessionStore, identityStore, prekeyStore, signedPrekeyStore, kyberPrekeyStore) {
    return Native.SessionCipher_DecryptPreKeySignalMessage(message, address, bridgeSessionStore(sessionStore), identityStore, bridgePreKeyStore(prekeyStore), bridgeSignedPreKeyStore(signedPrekeyStore), bridgeKyberPreKeyStore(kyberPrekeyStore));
}
export async function sealedSenderEncryptMessage(message, address, senderCert, sessionStore, identityStore) {
    const ciphertext = await signalEncrypt(message, address, sessionStore, identityStore);
    const usmc = UnidentifiedSenderMessageContent.new(ciphertext, senderCert, ContentHint.Default, null);
    return await sealedSenderEncrypt(usmc, address, identityStore);
}
export function sealedSenderEncrypt(content, address, identityStore) {
    return Native.SealedSender_Encrypt(address, content, identityStore);
}
export async function sealedSenderMultiRecipientEncrypt(contentOrOptions, recipients, identityStore, sessionStore) {
    let excludedRecipients = undefined;
    if (contentOrOptions instanceof UnidentifiedSenderMessageContent) {
        if (!recipients || !identityStore || !sessionStore) {
            throw Error('missing arguments for sealedSenderMultiRecipientEncrypt');
        }
    }
    else {
        ({
            content: contentOrOptions,
            recipients,
            excludedRecipients,
            identityStore,
            sessionStore,
        } = contentOrOptions);
    }
    const recipientSessions = await sessionStore.getExistingSessions(recipients);
    return await Native.SealedSender_MultiRecipientEncrypt(recipients, recipientSessions, ServiceId.toConcatenatedFixedWidthBinary(excludedRecipients ?? []), contentOrOptions, identityStore);
}
// For testing only
export function sealedSenderMultiRecipientMessageForSingleRecipient(message) {
    return Native.SealedSender_MultiRecipientMessageForSingleRecipient(message);
}
export async function sealedSenderDecryptMessage(message, trustRoot, timestamp, localE164, localUuid, localDeviceId, sessionStore, identityStore, prekeyStore, signedPrekeyStore, kyberPrekeyStore) {
    const ssdr = await Native.SealedSender_DecryptMessage(message, trustRoot, timestamp, localE164, localUuid, localDeviceId, bridgeSessionStore(sessionStore), identityStore, bridgePreKeyStore(prekeyStore), bridgeSignedPreKeyStore(signedPrekeyStore), bridgeKyberPreKeyStore(kyberPrekeyStore));
    return SealedSenderDecryptionResult._fromNativeHandle(ssdr);
}
export async function sealedSenderDecryptToUsmc(message, identityStore) {
    const usmc = await Native.SealedSender_DecryptToUsmc(message, identityStore);
    return UnidentifiedSenderMessageContent._fromNativeHandle(usmc);
}
export class Cds2Client {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static new(mrenclave, attestationMsg, currentTimestamp) {
        return new Cds2Client(Native.Cds2ClientState_New(mrenclave, attestationMsg, currentTimestamp.getTime()));
    }
    initialRequest() {
        return Native.SgxClientState_InitialRequest(this);
    }
    completeHandshake(buffer) {
        return Native.SgxClientState_CompleteHandshake(this, buffer);
    }
    establishedSend(buffer) {
        return Native.SgxClientState_EstablishedSend(this, buffer);
    }
    establishedRecv(buffer) {
        return Native.SgxClientState_EstablishedRecv(this, buffer);
    }
}
export class HsmEnclaveClient {
    constructor(nativeHandle) {
        this._nativeHandle = nativeHandle;
    }
    static new(public_key, code_hashes) {
        code_hashes.forEach((hash) => {
            if (hash.length != 32) {
                throw new Error('code hash length must be 32');
            }
        });
        const concat_hashes = Buffer.concat(code_hashes);
        return new HsmEnclaveClient(Native.HsmEnclaveClient_New(public_key, concat_hashes));
    }
    initialRequest() {
        return Native.HsmEnclaveClient_InitialRequest(this);
    }
    completeHandshake(buffer) {
        return Native.HsmEnclaveClient_CompleteHandshake(this, buffer);
    }
    establishedSend(buffer) {
        return Native.HsmEnclaveClient_EstablishedSend(this, buffer);
    }
    establishedRecv(buffer) {
        return Native.HsmEnclaveClient_EstablishedRecv(this, buffer);
    }
}
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["Error"] = 1] = "Error";
    LogLevel[LogLevel["Warn"] = 2] = "Warn";
    LogLevel[LogLevel["Info"] = 3] = "Info";
    LogLevel[LogLevel["Debug"] = 4] = "Debug";
    LogLevel[LogLevel["Trace"] = 5] = "Trace";
})(LogLevel || (LogLevel = {}));
export function initLogger(maxLevel, callback) {
    let nativeMaxLevel;
    switch (maxLevel) {
        case LogLevel.Error:
            nativeMaxLevel = 1 /* Native.LogLevel.Error */;
            break;
        case LogLevel.Warn:
            nativeMaxLevel = 2 /* Native.LogLevel.Warn */;
            break;
        case LogLevel.Info:
            nativeMaxLevel = 3 /* Native.LogLevel.Info */;
            break;
        case LogLevel.Debug:
            nativeMaxLevel = 4 /* Native.LogLevel.Debug */;
            break;
        case LogLevel.Trace:
            nativeMaxLevel = 5 /* Native.LogLevel.Trace */;
            break;
    }
    Native.initLogger(nativeMaxLevel, (nativeLevel, target, file, line, message) => {
        let level;
        switch (nativeLevel) {
            case 1 /* Native.LogLevel.Error */:
                level = LogLevel.Error;
                break;
            case 2 /* Native.LogLevel.Warn */:
                level = LogLevel.Warn;
                break;
            case 3 /* Native.LogLevel.Info */:
                level = LogLevel.Info;
                break;
            case 4 /* Native.LogLevel.Debug */:
                level = LogLevel.Debug;
                break;
            case 5 /* Native.LogLevel.Trace */:
                level = LogLevel.Trace;
                break;
            default:
                callback(LogLevel.Warn, 'signal-client', 'index.ts', 0, `unknown log level ${nativeLevel}; treating as error`);
                level = LogLevel.Error;
                break;
        }
        callback(level, target, file, line, message);
    });
}
